packages/core/src/routes/phrase.ts

```diff
@@ -0,0 +1,130 @@
+import assertThat from '@/utils/assert-that';
+
+import { AnonymousRouter } from './types';
+
+const getLanguageInfo = async (applicationId: unknown) => {
+  if (applicationId === adminConsoleApplicationId) {
+    return adminConsoleSignInExperience.languageInfo;
+  }
+
+  const { languageInfo } = await findDefaultSignInExperience();
+
+  return languageInfo;
+};
+
+const isBuiltInLanguageKey = (key: string): key is LanguageKey =>
+  Object.keys(resource).includes(key) && languageKeyGuard.safeParse(key).success;
+
+const isFullyTranslated = (fullTranslation: Translation, targetTranslation: Translation) => {
+  const keys = Object.keys(fullTranslation);
+  const targetKeys = new Set(Object.keys(targetTranslation));
+
+  if (keys.length !== targetKeys.size) {
+    return false;
+  }
+
+  if (!keys.every((key) => targetKeys.has(key) && notFalsy(targetTranslation[key]))) {
+    return false;
+  }
+
+  for (const key of keys) {
+    if (
+      typeof fullTranslation[key] === 'object' &&
+      // eslint-disable-next-line no-restricted-syntax
+      !isFullyTranslated(fullTranslation[key] as Translation, targetTranslation[key] as Translation)
+    ) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+const finalDefaultLanguage = 'en';
+// Ensure the final default language phrase contains a built-in phrase (that means it must be fully translated).
+assertThat(
+  Object.keys(resource).includes(finalDefaultLanguage),
+  new Error(`The final default language "${finalDefaultLanguage}" must contain a built-in phrase.`)
+);
+
+const getPhrase = async (
+  languageKey: string,
+  defaultLanguageKey = finalDefaultLanguage
+): Promise<ResourceLanguage> => {
+  const customLanguageKeys = await findAllCustomLanguageKeys();
+  const hasCustomPhrase = customLanguageKeys.includes(languageKey);
+  const hasBuiltInPhrase = isBuiltInLanguageKey(languageKey);
+
+  if (!hasCustomPhrase && !hasBuiltInPhrase) {
+    return getPhrase(defaultLanguageKey);
+  }
+
+  if (!hasCustomPhrase && hasBuiltInPhrase) {
+    return { languageKey, ...resource[languageKey] };
+  }
+
+  const customPhrase = await findCustomPhraseByLanguageKey(languageKey);
+
+  if (isFullyTranslated(en.translation, customPhrase.translation)) {
+    return customPhrase;
+  }
+
+  if (hasBuiltInPhrase) {
+    return deepmerge(resource[languageKey], customPhrase);
+  }
+
+  const defaultPhrase = await getPhrase(defaultLanguageKey);
+
+  return deepmerge(defaultPhrase, customPhrase);
+};
+
+export default function phraseRoutes<T extends AnonymousRouter>(router: T, provider: Provider) {
+  router.get(
+    '/phrase',
+    koaGuard({
+      response: CustomPhrases.guard,
+    }),
+    async (ctx, next) => {
+      const interaction = await provider
+        .interactionDetails(ctx.req, ctx.res)
+        .catch((error: unknown) => {
+          // Should not block if interaction is not found
+          if (error instanceof errors.SessionNotFound) {
+            return null;
```

Comment:
should not guard the session strictly? If no session is found return just return the default language settings.